### PR TITLE
OKTA-531964 Add form validation support.

### PR DESCRIPTION
**Pull request by JayNewstrom**

### CODE REVIEW

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Form.kt [hunk 1]
```diff
@@ -15,17 +15,75 @@
...
+    private val builderToElementMap: Map<Element.Builder<*>, Element>,
 ) {
     internal data class Builder(
         val elements: MutableList<Element.Builder<*>> = mutableListOf(),
         val launchActions: MutableList<suspend CoroutineScope.() -> Unit> = mutableListOf(),
     ) {
-        fun build(): Form {
-            return Form(elements.map { it.build() })
+        fun build(formTransformers: List<FormTransformer>): Form {
+            val builtFromWithoutTransform = Builder(
```
Comment (rajdeepnanua-okta) [inline native-authentication/src/main/java/com/okta/nativeauthentication/form/Form.kt:32]:
```suggestion
            val builtFormWithoutTransform = Builder(
```

**Review by rajdeepnanua-okta — COMMENTED** [review]
Had a couple questions

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt [hunk 2]
```diff
@@ -76,36 +136,91 @@ sealed interface Element {
...
+                    errorMessage = errorMessage,
+                    option = option,
+                )
+            }
+
+            override fun setValueFromElement(element: Options) {
+                option = element.option
+            }
+
+            override fun validate(remediation: IdxRemediation): Boolean {
```
[the remark this answers is not in the document]
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt:216]: This is used internally to only validate the form section (remediation) that was submitted. If they match, they're from the same section of the form, if they don't, it's from a different section, and we don't want to validate it.

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Form.kt (hunk 1 above)
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/form/Form.kt:32]: The was intentionally `builtFrom`. Basically marking the builder that was used to build the form.

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt [hunk 3]
```diff
@@ -17,18 +17,38 @@ package com.okta.nativeauthentication.form
...
     sealed class Builder<E : Element> {
         internal abstract fun build(): E
+
+        internal open fun setValueFromElement(element: E) {
+            // Value setting will happen in subclasses if necessary.
+        }
+
+        internal fun setValueFromBaseElement(element: Element) {
+            @Suppress("UNCHECKED_CAST")
+            setValueFromElement(element as E)
```
[the remark this answers is not in the document]
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt:30]: Yeah, this sort of pattern is pretty common. The JVM has type erasure, so you can't do an `element is E` check.

**Review by rajdeepnanua-okta — APPROVED** [review]

**Pull request merged**